pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/util/SegmentDeletionManagerTest.java

```diff
@@ -250,18 +274,83 @@ public void testRemoveDeletedSegments()
     // Check that dummy directories and files are successfully created.
     Assert.assertEquals(dummyDir1.list().length, 3);
     Assert.assertEquals(dummyDir2.list().length, 3);
+    Assert.assertEquals(dummyDir3.list().length, 3);
 
-    // Try to remove files with the retention of 3 days.
-    deletionManager.removeAgedDeletedSegments(3);
-    Assert.assertEquals(dummyDir1.list().length, 3);
-    Assert.assertEquals(dummyDir2.list().length, 1);
+    // Try to remove files with the retention of 1 days.
+    deletionManager.removeAgedDeletedSegments();
 
-    // Try to further remove files with the retention of 1 days.
-    deletionManager.removeAgedDeletedSegments(1);
+    // Check that only 1 day retention file is remaining
     Assert.assertEquals(dummyDir1.list().length, 1);
 
     // Check that empty directory has successfully been removed.
     Assert.assertEquals(dummyDir2.exists(), false);
+
+    // Check that deleted file without retention suffix is honoring cluster-wide retention period of 7 days.
+    Assert.assertEquals(dummyDir3.list().length, 1);
+  }
+
+  @Test
+  public void testSegmentDeletionLogic()
+      throws Exception {
+    Map<String, Object> properties = new HashMap<>();
+    properties.put(CommonConstants.Controller.PREFIX_OF_CONFIG_OF_PINOT_FS_FACTORY + ".class",
+        LocalPinotFS.class.getName());
+    PinotFSFactory.init(new PinotConfiguration(properties));
+
+    HelixAdmin helixAdmin = makeHelixAdmin();
+    ZkHelixPropertyStore<ZNRecord> propertyStore = makePropertyStore();
+    File tempDir = Files.createTempDir();
+    tempDir.deleteOnExit();
+    SegmentDeletionManager deletionManager = new SegmentDeletionManager(
+        tempDir.getAbsolutePath(), helixAdmin, CLUSTER_NAME, propertyStore, 7);
+
+    // create table segment files.
+    Set<String> segments = new HashSet<>(segmentsThatShouldBeDeleted());
+    createTableAndSegmentFiles(tempDir, segmentsThatShouldBeDeleted());
+    File tableDir = new File(tempDir.getAbsolutePath() + File.separator + TABLE_NAME);
+    File deletedTableDir = new File(tempDir.getAbsolutePath() + File.separator + "Deleted_Segments"
+        + File.separator + TABLE_NAME);
+
+    // delete the segments instantly.
+    SegmentsValidationAndRetentionConfig mockValidationConfig = mock(SegmentsValidationAndRetentionConfig.class);
+    when(mockValidationConfig.getDeletedSegmentsRetentionPeriod()).thenReturn("0d");
+    TableConfig mockTableConfig = mock(TableConfig.class);
+    when(mockTableConfig.getValidationConfig()).thenReturn(mockValidationConfig);
+    deletionManager.deleteSegments(TABLE_NAME, segments, mockTableConfig);
+
+    // Sleep 3 second to ensure the async delete actually kicked in.
+    Thread.sleep(3000L);
```

Comment:
This can potentially be flaky. Suggest using `TestUtils.waitForCondition()` with a longer timeout